src/transformers/modeling_fsmt.py

```diff
@@ -0,0 +1,1341 @@
+        # mfairseqtranslator has one extra layer_norm
+        self.layer_norm = LayerNorm(config.d_model) if config.normalize_before else None
+
+    def forward(
+        self, input_ids, attention_mask=None, output_attentions=False, output_hidden_states=False, return_dict=False
+    ):
+        """
+        Args:
+            input_ids (LongTensor): tokens in the source language of shape
+                `(batch, src_len)`
+            attention_mask (torch.LongTensor): indicating which indices are padding tokens.
+        Returns:
+            BaseModelOutput or Tuple comprised of:
+                - **x** (Tensor): the last encoder layer's output of
+                  shape `(src_len, batch, embed_dim)`
+                - **encoder_states** (tuple(torch.FloatTensor)): all intermediate
+                  hidden states of shape `(src_len, batch, embed_dim)`.
+                  Only populated if *output_hidden_states:* is True.
+                - **all_attentions** (tuple(torch.FloatTensor)): Attention weights for each layer.
+                During training might not be of length n_layers because of layer dropout.
+        """
+        # check attention mask and invert
+        if attention_mask is not None:
+            attention_mask = invert_mask(attention_mask)
+
+        inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale
+        embed_pos = self.embed_positions(input_ids)
+        x = inputs_embeds + embed_pos
+        x = self.layernorm_embedding(x)
+        x = F.dropout(x, p=self.dropout, training=self.training)
+
+        # B x T x C -> T x B x C
+        x = x.transpose(0, 1)
+
+        encoder_states = [] if output_hidden_states else None
+        all_attentions = () if output_attentions else None
+        for encoder_layer in self.layers:
+            if output_hidden_states:
+                encoder_states.append(x)
+            # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
+            dropout_probability = random.uniform(0, 1)
+            if self.training and (dropout_probability < self.layerdrop):  # skip the layer
+                attn = None
+            else:
+                x, attn = encoder_layer(x, attention_mask, output_attentions=output_attentions)
+
+            if output_attentions:
+                all_attentions = all_attentions + (attn,)
+
+        if self.layer_norm:
+            x = self.layer_norm(x)
+        if output_hidden_states:
+            encoder_states.append(x)
+            # T x B x C -> B x T x C
+            encoder_states = tuple(hidden_state.transpose(0, 1) for hidden_state in encoder_states)
+
+        # T x B x C -> B x T x C
+        x = x.transpose(0, 1)
+
+        if not return_dict:
+            return tuple(v for v in [x, encoder_states, all_attentions] if v is not None)
+        return BaseModelOutput(last_hidden_state=x, hidden_states=encoder_states, attentions=all_attentions)
+
+
+class DecoderLayer(nn.Module):
```

Comment:
Can you import helper functions and classes that are **identical** to bart? I think that's allowed :)

Reply:
Great idea - was meaning to compare if anything is re-usable: salvaged as is `Attention, EncoderLayer, DecoderLayer`

Currently, it's not great, since there is a lot of code catering to Bart's subclasses, so eventually it either needs to be removed if this dual dict is going into a new group of models, or somehow merged back into Bart.

But this is a start...

Reply:
Hahaha no that's not allowed @sshleifer 

You are only allowed to import from another model if the **full model** is identical (and probably you should just have a checkpoint and not add a model to the library in this case).

Can you guys @sshleifer @LysandreJik @stas00 ping me to review this PR before merging?

Reply:
Makes sense, ok!